README.md
![](vignettes/gs_browse1.png)

# Introduction
A tutorial overview of `flowPloidy` is available on
the
[Bioconductor website](http://bioconductor.org/packages/release/bioc/vignettes/flowPloidy/inst/doc/flowPloidy-gettingStarted.html).
This vignette is provided with the package, so once you have `flowPloidy`
installed you can access it from with R (see below).

# Installation

## Stable Version

`flowPloidy` is available in [Bioconductor](https://bioconductor.org).

To install it, you need to install the `bioconductor` R package (more
details on the [Bioconductor site ](http://bioconductor.org/install/)):

```{r}
## try http:// if https:// URLs are not supported
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install()
```

Once that's installed, you can install `flowPloidy` using the Bioconductor
tools:

```{r}
BiocManager::install("flowPloidy")
BiocManager::install("flowPloidyData")   # (optional) data for the examples
```

This should pull in all the package dependencies for `flowPloidy`, after
which you can load the package with the normal function
`library("flowPloidy")`.

## Development Version

As of June 2018, I have added a new analysis method. This is aimed at
assessing endopolyploidy, where a single sample may have four or more
peaks. The intent is to compare the number of cells in each peak, rather
than to determine a ratio relative to a co-chopped standard.

This new code will be incorporated into Bioconductor for the next release.
If you'd like to try it now, you can install it directly from the GitHub
repository as follows: 

```{r}
## Install Bioconductor tools first:
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install()

## Install flowCore from Bioconductor:
BiocManager::install("flowCore")

## Install devtools so you can directly access GitHub
install.packages(devtools)
library(devtools)

## Install flowPloidy:
install_github("plantarum/flowPloidy", dependencies = TRUE, 
    build_vignettes = TRUE)
```

If the last command fails, particularly with complaints about building a
vignette, or reference to Pandoc, try with `build_vignettes = FALSE`
instead.

Note that I haven't yet updated the documentation to match the new code. To
use the endopolyploidy workflow, you need to use a new argument, `g2 =
FALSE` in your call to `FlowHist` or `batchFlowHist` (**NB: use g2, lowercase, not G2, uppercase. The original version of this README was incorrect!**). This excludes the g2
peaks from peak fitting, treating each peak as an independent group of
cells. You may also want to increase the `samples` argument to match the
number of peaks; however, you can correct this in `browseFlowHist`, so
that's not critical.


```{r }
## loading files for endopolyploidy analysis:
batch1 <- batchFlowHist(endo_files, channel="FL3.INT.LIN", g2 = FALSE,
    samples = 5)

batch1 <- browseFlowHist(batch1)
```

Expanding `flowPloidy` to handle up to six peaks (and now potentially an
unlimited number if needed) required reworking a bunch of the existing
code, and as part of this the column headings in the tables produced by
`tabulateFlowHist` are now different from the previous release.

# Getting Started

```{r}
library("flowPloidy")
```

The `flowPloidy` workflow is documented in the vignette, which you can view
from R:

```{r}
fpVig <- vignette("flowPloidy-overview")
fpVig ## open vignette in a browser
edit(name = fpVig) ## open vignette source code in a text editor
```

It is also available [online](http://bioconductor.org/packages/release/bioc/vignettes/flowPloidy/inst/doc/flowPloidy-gettingStarted.html).

# Getting Help
For general help using the package, you can post questions on
the [Bioconductor Support Site](https://support.bioconductor.org/). Use the
tag `flowploidy` to ensure your question is brought to my attention.

The development repository for `flowPloidy` is
on [Github](https://github.com/plantarum/flowPloidy), and you can file bugs
there using the **issues** tab. You are also welcome to contribute features
or bug-fixes via pull requests!


inst/NEWS
Changes in version 1.7.4 (2019-04-01)

User Visible Changes:

    * Note that the data files in `flowPloidyData` are now accessed by
      functions, not variables. That is, instead of refering to the
      variable `flowPloidyFiles`, you now need to refer to the output of
      the function `flowPloidyFiles()`. This was necessary to accomodate
      changes in the way R loads packages. This only effects the sample
      data from `flowPloidyData`, you should not change code referring to
      your own files.

Changes in version 1.7.2 (2018-10-03)
------------------------------------

WARNING:

    * There are some major user-visible changes in this version. The
      vignette has been updated, as have the manual pages. There may still
      be inconsistencies, please let me know if you find anything
      confusing. I still need to document the endopolyploidy workflow.

User Visible Changes:

    * Column names reported in tabulateFlowHist are completely revised. The
      new names should be self-explanatory, but will not match the column
      names reported in previous versions of flowPloidy. The output now
      includes the counts for the G1 and G2 peaks, as well as for each
      S-phase component.

    * The browseFlowHist interface now has a "fail" button. This provides
      the user with the ability to reject a sample, even if flowPloidy can
      fit a model to it. For situations where flowPloidy succeeds in
      fitting a model to noise, or where only your standard is visible;
      selecting "fail" for these samples will guarantee that they show up
      as NAs in your summary data.

    * The browseFlowHist interface now includes the option to add
      annotations. These are text comments, not part of the analysis. This
      allows the user to add notes such as "this sample needs to be rerun",
      or "exclude this, data is poor quality", etc.

    * Analysis of endopolyploidy is supported by allowing the user to a)
      increase the number of peaks (samples) fit, up to a maximum of 6; b)
      to *not* fit G2 peaks (via the G2 = FALSE argument), so that the 2C,
      4C, 6C etc peaks are all fit independently of each other, without
      requiring that the 4C peak be twice the size of the 2C peak; and c)
      reworking the output of tabulateFlowHist to allow for reporting the
      fit parameters for an arbitrary number of peaks.

    * Internal changes to the model component code uncovered some bugs in
      the way G2 peaks were fit. These are fixed, and the model fitting
      works slightly better (higher success rate) as a consequence.

Internal Changes:

    * The G1, G2 and S-phase model components are now generated by helper
      functions. These allows me to tweak the component definitions in a
      single location, after which the change is automatically incorporated
      in all of the associated components.

    * The above changes allowed me to replace a lot of duplicated code with
      more general forms, notably in the plotting and tabulating functions.

    * Model parameters are now collected dynamically, so I don't need to
      hard-code the output of tabulateFlowHist. Parameters that the user
      won't need are indicated with the suffix 'P'; any parameter ending
      with 'P' is excluded from the output.

    * Added a 'fail' slot to the FlowHist class, to support the fail button
      described above.

Changes in version 1.5.6 (2018-03-08)
-------------------------------------

User Visible Changes:

    * BUG FIX: the G2 peak of the B sample was not getting incorporated
      into model construction, which caused model fitting to fail on
      samples with histograms skewed towards the left.

    * Updated DebrisModel documentation.

Internal Changes:

    * Fixed broken test.

Changes in version 1.5.3 (2018-01-17)
-------------------------------------

User Visible Changes:

    * The browser interface will no longer allow users to enter arbitrary
      text to select the number of samples. Only valid values, i.e., 1, 2,
      or 3, will be offered as choices. The old version would crash with a
      bad value for sample number.

    * Improved file importing, so that low-quality samples for which peaks
      cannot be detected can still be imported.

    * Added a new argument to FlowHist, so users can set the threshold
      below which the data is ignored when screening out debris. The
      default remains the same, 40, but users with very clean histograms,
      with peaks far to the left, can now lower this value if needed.

    * Model fitting is now limited to the range of the data. That is, empty
      bins at the right (upper) end of the histogram will not be included
      when fitting the NLS. This addresses problems that generated inflated
      RCS values.

    * The limits for the linearity parameter have been extended from
      1.9-2.1, to 1.5-2.5. This will improve/enable model fitting on
      samples where the linearity (the ratio of the G2/G1 peaks) was outside
      the range 1.9-2.1.

Changes in version 1.5.1 (2017-12-04)
-------------------------------------

Minor bug squashed:

    * Changes made upstream to `car::deltaMethod()` introduced a bug. This
      will be resolved as of `car` version 2.1-7. Until that version of
      `car` makes it's way into CRAN, calling `deltaMethod()` on an `nls`

tests/testthat.R
library(testthat)
library(flowPloidy)

test_check("flowPloidy")


tests/testthat/testcore.R
library(flowPloidy)
library(flowPloidyData)
context("core features")

fh1 <- FlowHist(file = flowPloidyFiles()["188-15.LMD"],
                channel = "FL3.INT.LIN", analyze = TRUE)

fh2 <- FlowHist(file = flowPloidyFiles()["222.LMD"],
                channel = "FL3.INT.LIN", analyze = FALSE)

fh3 <- FlowHist(file = flowPloidyFiles()["240+S.LMD"],
                channel = "FL3.INT.LIN", debris = "MC", analyze = TRUE)

test_that("FlowHist objects print without error", {
  expect_error(print(fh1), NA)
  expect_error(print(fh2), NA)
  expect_error(print(fh3), NA)
})

## Testing that analyzed, unanalyzed, and groups of possibly mixed analyzed
## and unanalyzed FlowHist objects are all handled politely by
## tabulateFlowHist:  
test_that("FlowHist objects tabulate without error", {
  expect_error(tabulateFlowHist(fh1), NA)
  expect_error(tabulateFlowHist(fh2), NA)
  expect_error(tabulateFlowHist(list(fh1, fh2), NA))
  expect_error(tabulateFlowHist(list(fh2, fh3)), NA)
  expect_error(tabulateFlowHist(list(fh1, fh2, fh3)), NA)
})

test_that("FlowHist plots work without error", {
  expect_error(plot(fh1, init = TRUE), NA)
  expect_error(plot(fh2, init = TRUE), NA)
  expect_error(plot(fh3, init = TRUE), NA)
})

test_that("Crappy FCM files can be loaded and plotted without error", {
  expect_error(fhBad <- FlowHist(fpBad(), channel = "FL2.A"), NA)
  expect_error(plot(fhBad, init = TRUE), NA)
  expect_error(print(fhBad), NA)
})


tests/testthat/testmodels.R
library(flowPloidy)
library(flowPloidyData)
context("models and components")

fh1 <- FlowHist(file = flowPloidyFiles()["188-15.LMD"],
                channel = "FL3.INT.LIN")
fh2 <- fh1

## ignore peak at 100, such that both A and B components have no G2 peak
fhPeaks(fh2)[1, 1] <- fhPeaks(fh2)[1, 1] * 2
fh2 <- updateFlowHist(fh2)

## replace peak at 100, A component should now have G1 and G2 peak
fh3 <- fh2

fhPeaks(fh3)[1, 1] <- fhPeaks(fh3)[1, 1] / 2
fh3 <- addComponents(fh3)
fh3 <- makeModel(fh3)
fh3 <- getInit(fh3)

test_that("Peak components are updated when selecting peaks", {
  expect_true("a1" %in% names(fhComps(fh1)))
  expect_true("a2" %in% names(fhComps(fh1)))
  expect_true("b1" %in% names(fhComps(fh1)))
  expect_false("b2" %in% names(fhComps(fh1)))

  expect_true("a1" %in% names(fhComps(fh2)))
  expect_false("a2" %in% names(fhComps(fh2)))
  expect_true("b1" %in% names(fhComps(fh2)))
  expect_false("b2" %in% names(fhComps(fh2)))

  expect_true("a1" %in% names(fhComps(fh3)))
  expect_true("a2" %in% names(fhComps(fh3)))
  expect_true("b1" %in% names(fhComps(fh3)))
  expect_false("b2" %in% names(fhComps(fh3)))
})



R/flowGate.R
## Initially this was a much larger file containing a number of approaches
## to gating. Now that we've settled on one, it doesn't actually require
## much code. We could probably move this back into FlowHist.R

## This function isn't actually used anymore in FlowHist. It might be
## useful for providing users with a way to visualize gates from the
## console? Or at least, could be extended/repurposed for that use?  
plotGate <- function(fh, x, y, ...){
  chans <- viewFlowChannels(fh)
  if(! x %in% chans)
    stop("Selected flow channel: ", x, " does not exist!")
  if(! y %in% chans)
      stop("Selected flow channel: ", y, " does not exist!")

  dat <- exprs(fhRaw(fh))[, c(x, y)]
  dat[,y] <- dat[,y] / dat[,x]
  plot(dat, pch = 16, col = "#05050510")
}

## Called from the shiny interface
plotResid <- function(fh, main = fhFile(fh), sub = "Gate Residuals", ...){
  plot(fhHistData(fh)$gateResid, type = 'n', main = main,
       sub = sub, xlab = fhChannel(fh), ylab = "Intensity",
       ...) 
  polygon(x = c(fhHistData(fh)$xx, max(fhHistData(fh)$xx), 0), 
          y = c(fhHistData(fh)$gateResid, 0, 0),
          col = "lightgray", border = NA)
}

#' Apply a gate to a FlowHist object
#'
#' This function is primarily book-keeping to make sure that
#' \code{histData} and downstream calculations are appropriately updated
#' when a gate is applied. The code for applying the gate is actually in
#' the function \code{\link{setBins}}.
#' 
#' @title setGate
#' @param fh a \code{\link{FlowHist}} object
#' @param gate boolean, a vector indicating which rows in the raw data
#'   should be included (gated) in the analysis.
#' @param refresh boolean, should the analysis be updated after applying
#'   the gate (default = TRUE)?
#' @return \code{setGate} returns an updated \code{\link{FlowHist}} object,
#'   with the \code{histData} slot updated to account for the gate. With
#'   \code{refresh = TRUE} (default), it will also rebuild the model and
#'   complete the analysis.
#'
#'   \code{isGated} returns TRUE if the \code{\link{FlowHist}} object is
#'   gated.
#' @author Tyler Smith
#' @seealso \code{\link{setBins}}
#' @keywords internal
setGate <- function(fh, gate, refresh = TRUE){
  fhGate(fh) <- gate
  ## We save and restore the peaks here. In most cases, you need to
  ## relocate peaks after setBins, since you're changing the underlying
  ## data the peaks are located in. When setting a gate, however, the peak
  ## position is often/usually going to remain in the same place, the gate
  ## only removes marginal noisy areas (debris). setBins could be modified
  ## to leave the peaks unaltered, but that might lead to abuse in
  ## situations where we want to enforce re-finding peaks.
  peaks <- fhPeaks(fh)
  fh <- setBins(fh, fhBins(fh))
  fhPeaks(fh) <- peaks
  if(refresh){
    fh <- addComponents(fh)
    fh <- setLimits(fh)
    fh <- makeModel(fh)
    fh <- getInit(fh)
    fh <- fhAnalyze(fh)
  }
  fh
}

#' @rdname setGate
isGated <- function(fh){
  ## returns TRUE if the FlowHist data is gated
  sum(fhGate(fh)) != 0
}



R/shiny.R
#' @importFrom shiny actionButton brushOpts brushedPoints column
#'   eventReactive eventReactive fluidPage fluidRow HTML htmlOutput
#'   mainPanel nearPoints numericInput observe observeEvent plotOutput
#'   radioButtons reactive reactiveValues renderPlot renderPrint
#'   renderTable renderText runApp selectInput shinyApp sidebarLayout
#'   sidebarPanel sliderInput stopApp tags titlePanel updateNumericInput
#'   updateRadioButtons updateSelectInput updateSliderInput wellPanel
#'   tabsetPanel tabPanel textInput updateTextInput
NULL

#' @importFrom utils str
NULL

#' @importFrom methods is
NULL

#' Visually assess and correct histogram fits
#'
#' Visually assess histogram fits, correcting initial values, and selecting
#' model components.
#'
#' This function will open a browser tab displaying the first
#' \code{\link{FlowHist}} object from the argument \code{flowList}. Using
#' the interface, the user can modify the starting values for the histogram
#' peaks, select different debris model components, toggle the linearity
#' option, select which peak to treat as the standard, and, if multiple
#' standard sizes are available, select which one to apply.
#'
#' See the "Getting Started" vignette for a tutorial introduction.
#' 
#' @title browseFlowHist
#' @param flowList either a \code{\link{FlowHist}} object, or a list of
#'   \code{\link{FlowHist}} objects
#' @param debug boolean, turns on debugging messages
#' @return Returns the list of \code{\link{FlowHist}} objects, updated by
#'   any changes made in the GUI.
#' @author Tyler Smith
#' @examples
#' library(flowPloidyData)
#' batch1 <- batchFlowHist(flowPloidyFiles(), channel = "FL3.INT.LIN")
#' \dontrun{
#' batch1 <- browseFlowHist(batch1)
#' }
#' @export
browseFlowHist <- function(flowList, debug = FALSE){
  if(is(flowList, "FlowHist")){
    ## if flowList is a single FlowHist object, wrap it in a list so we can
    ## use the same code for processing single and multiple FlowHist
    ## objects. Convert back to a single object on return.
    flowList <- list(flowList)
    names(flowList) <- fhFile(flowList[[1]])
  }
  
  .fhI <- 1
  .fhList <- flowList

  initialLinearity <- fhLinearity(.fhList[[1]])

  if(debug) message("init Linearity: ", initialLinearity)

  initialStdSelected <- fhStdSelected(.fhList[[1]])
  if(debug) message("initialStdSelected: ", initialStdSelected)
  standardList <- fhStdSizes(.fhList[[1]])
  if(debug) message("standardList: ", standardList)
  initialStdPeak <- fhStdPeak(.fhList[[1]])
  
  initialDebris <- fhDebris(.fhList[[1]])

  initialSamples <- fhSamples(.fhList[[1]])

  initialAnnotation <- fhAnnotation(.fhList[[1]])

  initialFail <- fhFail(.fhList[[1]])
  
  if(debug) message("init Debris: ", initialDebris)

  raw <- exprs(fhRaw(.fhList[[.fhI]]))
  chan1 <- fhChannel(.fhList[[.fhI]])
  if(length(viewFlowChannels(.fhList[[.fhI]])) > 1){
    ## why 2? sometimes, by chance, the second column is the SS value.
    chan2 <- viewFlowChannels(.fhList[[.fhI]])[2]
  } else {
    ## there's only one channel, so gating is pointless
    chan2 <- chan1
  }
  
  initGateData <- data.frame(x = raw[, chan1],
                             y = raw[, chan2] / raw[, chan1]) 

  maxY <- max(initGateData$y[!is.infinite(initGateData$y)],
              na.rm = TRUE)

  if(is.nan(maxY))
    maxY <- 6
  
  ui <- fluidPage(
    tags$head(
           tags$style(HTML("
      .sidepanel {
          max-width: 300px;
      }
      #gatePlot, #gatedData, #gateResiduals {
          max-height: 300px;
      }
      #exit {
          margin-top: 1.4em;
      }
      #setGate {
          margin-top: 1.8em;
      }
    "))),
    fluidRow(
      column(width = 3,
             tags$div(class = "sidepanel", 
             wellPanel(
               fluidRow(
                 column(4,
                        actionButton("exit", label = "Exit")),
                 column(8, 
                        fluidRow(
                          column(12, htmlOutput("flowNumber",
                                                align = "center"))),
                        fluidRow(
                          column(5, actionButton("prev", label = "Prev")),
                          column(7, actionButton("nxt",
                                                 label = "Next"))))),
##               tags$hr(),
               fluidRow(
                 column(6,
                        ## If we use numericInput, the user is able to
                        ## enter invalid values in the textbox. With
                        ## selectInput we constrain the possible input
                        ## values to those that won't break anything.
                        selectInput('sampSelect', 'Samples',
                                    initialSamples,
                                    choices = list("1" = 1,
                                                   "2" = 2,
                                                   "3" = 3,
                                                   "4" = 4,
                                                   "5" = 5,
                                                   "6" = 6))), 
                 column(6,
                        selectInput(inputId = "peakPicker",
                                    label = "Peak",
                                    selected = "A",
                                    choices = list("A", "B", "C", "D", "E",
                                                   "F")))),
               fluidRow(
                 column(6, 
                        selectInput(inputId = 'standardSelect',
                                    label = 'Standard Value',
                                    choices = standardList,
                                    selected = initialStdSelected)),
                 column(6,
                        selectInput(inputId = "standardPeak",
                                    label = "Standard Peak",
                                    selected = initialStdPeak,
                                    choices = list("X", "A", "B")))), 
               fluidRow(
                 column(6,
                        selectInput(inputId = "linearity",
                                    label = "Linearity",
                                    selected = initialLinearity,
                                    choices = list("Fixed" = "fixed",
                                                   "Variable" =
                                                     "variable"))),  
                 column(6, 
                        selectInput(inputId = "debris",
                                    label = "Debris Model",
                                    choices = list("MC" = "MC",
                                                   "SC" = "SC", 
                                                   "none" = "none"),  
                                    selected = initialDebris))),
               fluidRow(
                 column(6,
                        radioButtons(inputId = "failButton",
                                     label = "Sample Quality",
                                     choiceNames = c("Pass", "Fail"),
                                     choiceValues = c(FALSE, TRUE),
                                     selected = initialFail,
                                     inline = TRUE)))
             ))),
      column(width = 9,
             plotOutput("fhHistogram", click = "pointPicker"))
    ),
    tabsetPanel(type = "tabs",
      tabPanel("Annotation",
               textInput("annotation", "",
                         initialAnnotation, width = "100%"), 
               actionButton("updateAnnotation", label = "Annotate")),
      tabPanel("Gating", 
    fluidRow(
      column(width = 3,
             tags$div(class = "sidepanel",
             wellPanel(
               fluidRow(
                 column(6,
                        selectInput('xcol', 'X Variable',
                                    viewFlowChannels(.fhList[[.fhI]]), 
                                    selected = chan1)),
                 column(6,
                        selectInput('ycol', 'Y Variable',
                                    viewFlowChannels(.fhList[[.fhI]]), 
                                    selected = chan2))),
               sliderInput("yrange", "Zoom", min = 0, ticks = FALSE,
                           step =
                             max(6, ceiling(log(maxY)))/20, 
                           max = max(6, ceiling(log(maxY)), na.rm = TRUE),
                           value = 0, dragRange = FALSE),
               fluidRow(
                 column(6, 
                        selectInput('yType', 'Y axis', c("Y/X", "Y"),
                                    selected = "Y/X")),
                 column(6,
                        actionButton("setGate", label = "Set Gate"))
               )))),
      column(width = 3,
             plotOutput("gatePlot",
                        click = "gatePlot_click",
                        brush = brushOpts(id = "gatePlot_brush",
                                          resetOnNew = FALSE
                                          )
                        )
             ),
      column(width = 3,
             plotOutput("gatedData")
             ),
      column(width = 3,
             plotOutput("gateResiduals")))
    )
    )
    
  )
    
  server <- function(input, output, session){
    rv <- reactiveValues(fhI = 1, FH = .fhList[[1]])

    ## eventReactive would require an invalid reactive value in the
    ## expression to trigger the calculation; observeEvent will simply do
    ## the calculation:
    fhNext <- observeEvent(input$nxt, {
      if(rv$fhI < length(.fhList))
        rv$fhI <- rv$fhI + 1
    })

    fhPrev <- observeEvent(input$prev, {
      if(rv$fhI > 1)
        rv$fhI <- rv$fhI - 1
    })

    fhCurrent <- eventReactive(rv$fhI, {
      ## When navigating to a new FlowHist object via Prev/Next, update the
      ## radio buttons before updating rv$FH, so the replotting isn't
      ## triggered until the Radio Buttons are set to the values of the
      ## current FH object. The reaction chain is:

      ## fhNext/fhPrev --> rv$fhI --> fhCurrent --> rv$FH
      
      updateSelectInput(session, "linearity",
                         selected = fhLinearity(.fhList[[rv$fhI]]))
      updateSelectInput(session, "debris",
                         selected = fhDebris(.fhList[[rv$fhI]]))
      updateNumericInput(session, "sampSelect",
                         value = fhSamples(.fhList[[rv$fhI]]))

      updateSelectInput(session, "standardSelect",
                         selected = fhStdSelected(.fhList[[rv$fhI]]))
      updateSelectInput(session, "standardPeak",
                        selected = fhStdPeak(.fhList[[rv$fhI]]))
      updateRadioButtons(session, "failButton",
                        selected = fhFail(.fhList[[rv$fhI]]))
      updateTextInput(session, "annotation",
                      value = fhAnnotation(.fhList[[rv$fhI]]))
      rv$FH <- .fhList[[rv$fhI]]
      rv$fhI
    })      

    fhSetGate <- observeEvent(input$setGate, { 
      dat <- gateData()
      bp <- brushedPoints(dat, xvar = names(dat)[1],
                          yvar = names(dat)[2], input$gatePlot_brush,
                          allRows = TRUE)$selected_
      if(sum(bp) == 0) bp <- logical()
      
      .fhList[[fhCurrent()]] <<- setGate(.fhList[[fhCurrent()]], bp)
      rv$FH <- .fhList[[fhCurrent()]]
    })

    fhUpdateAnnotation <- observeEvent(input$updateAnnotation, {
      fhAnnotation(.fhList[[fhCurrent()]]) <<- input$annotation
    })
    
    fhPickPeaks <- observeEvent(input$pointPicker, {
      xPt <- nearPoints(fhHistData(.fhList[[fhCurrent()]]),
                        input$pointPicker, "xx", "intensity",
                        threshold = 25, maxpoints = 1)
      if(nrow(xPt) > 0){
        if(input$peakPicker == "A"){
          .fhList[[fhCurrent()]] <<-
            selectPeaks(.fhList[[fhCurrent()]],
                        xPt[1,1],
                        fhInit(.fhList[[fhCurrent()]])$b_mean,
                        fhInit(.fhList[[fhCurrent()]])$c_mean) 
          .fhList[[fhCurrent()]] <<- fhAnalyze(.fhList[[fhCurrent()]])
        } else if(input$peakPicker == "B"){
          .fhList[[fhCurrent()]] <<-
            selectPeaks(.fhList[[fhCurrent()]],
                        fhInit(.fhList[[fhCurrent()]])$a_mean,
                        xPt[1,1],
                        fhInit(.fhList[[fhCurrent()]])$c_mean)
          .fhList[[fhCurrent()]] <<- fhAnalyze(.fhList[[fhCurrent()]])
        } else if(input$peakPicker == "C"){
          .fhList[[fhCurrent()]] <<-
            selectPeaks(.fhList[[fhCurrent()]],
                        fhInit(.fhList[[fhCurrent()]])$a_mean,
                        fhInit(.fhList[[fhCurrent()]])$b_mean,
                        xPt[1,1])
          .fhList[[fhCurrent()]] <<- fhAnalyze(.fhList[[fhCurrent()]])
        } else if(input$peakPicker == "D"){
          .fhList[[fhCurrent()]] <<-
            selectPeaks(.fhList[[fhCurrent()]],
                        fhInit(.fhList[[fhCurrent()]])$a_mean,
                        fhInit(.fhList[[fhCurrent()]])$b_mean,
                        fhInit(.fhList[[fhCurrent()]])$c_mean,
                        xPt[1,1])
          .fhList[[fhCurrent()]] <<- fhAnalyze(.fhList[[fhCurrent()]])
        } else if(input$peakPicker == "E"){
          .fhList[[fhCurrent()]] <<-
            selectPeaks(.fhList[[fhCurrent()]],
                        fhInit(.fhList[[fhCurrent()]])$a_mean,
                        fhInit(.fhList[[fhCurrent()]])$b_mean,
                        fhInit(.fhList[[fhCurrent()]])$c_mean,
                        fhInit(.fhList[[fhCurrent()]])$d_mean,
                        xPt[1,1])
          .fhList[[fhCurrent()]] <<- fhAnalyze(.fhList[[fhCurrent()]])
        } else if(input$peakPicker == "F"){
          .fhList[[fhCurrent()]] <<-
            selectPeaks(.fhList[[fhCurrent()]],
                        fhInit(.fhList[[fhCurrent()]])$a_mean,
                        fhInit(.fhList[[fhCurrent()]])$b_mean,
                        fhInit(.fhList[[fhCurrent()]])$c_mean,
                        fhInit(.fhList[[fhCurrent()]])$d_mean,
                        fhInit(.fhList[[fhCurrent()]])$e_mean,
                        xPt[1,1])
          .fhList[[fhCurrent()]] <<- fhAnalyze(.fhList[[fhCurrent()]])
        }

        rv$FH <- .fhList[[fhCurrent()]]
        ## Changing peaks may change the linearity setting, need to update
        ## the picker:
        updateSelectInput(session, "linearity",
                          selected = fhLinearity(.fhList[[rv$fhI]]))
      }
    })

    ## Not sure why the following toggle events don't respond as
    ## eventReactives?
    fhToggleLinearity <- observeEvent(input$linearity, {
      if(fhLinearity(.fhList[[fhCurrent()]]) != input$linearity){
        .fhList[[fhCurrent()]] <<-
          updateFlowHist(.fhList[[fhCurrent()]],
                         linearity = input$linearity, analyze = TRUE)
        rv$FH <- .fhList[[fhCurrent()]]
      }
    })
    
    fhToggleDebris <- observeEvent(input$debris, {
      if(fhDebris(.fhList[[fhCurrent()]]) != input$debris){
        .fhList[[fhCurrent()]] <<-
          updateFlowHist(.fhList[[fhCurrent()]],
                         debris = input$debris, analyze = TRUE)
        rv$FH <- .fhList[[fhCurrent()]]
      }
    })

    fhToggleSamples <- observeEvent(input$sampSelect, {
      if(fhSamples(.fhList[[fhCurrent()]]) != input$sampSelect){
        .fhList[[fhCurrent()]] <<-
          updateFlowHist(.fhList[[fhCurrent()]],
                         samples = input$sampSelect, analyze = TRUE)
        rv$FH <- .fhList[[fhCurrent()]]
      }
    })

    fhUpdateFail <- observeEvent(input$failButton, {
      if(input$failButton){
        .fhList[[fhCurrent()]] <<- failFlowHist(.fhList[[fhCurrent()]])
        rv$FH <- .fhList[[fhCurrent()]]
      } else {
        .fhList[[fhCurrent()]] <<- passFlowHist(.fhList[[fhCurrent()]])
        rv$FH <- .fhList[[fhCurrent()]]
      }
    })

    fhUpdateStdPeak <- observeEvent(input$standardPeak, {
      if(fhStdPeak(.fhList[[fhCurrent()]]) != input$standardPeak){
        fhStdPeak(.fhList[[fhCurrent()]]) <<- input$standardPeak
        rv$FH <- .fhList[[fhCurrent()]]
      }
    })

    fhUpdateStdSelected <- observeEvent(input$standardSelect, {
      ## don't update the fh object if the input is the same as the actual
      ## value (may come up when switching to a new object)
      if(fhStdSelected(.fhList[[fhCurrent()]]) != input$standardSelect){
        fhStdSelected(.fhList[[fhCurrent()]]) <<-
          as.numeric(input$standardSelect)
        rv$FH <- .fhList[[fhCurrent()]]
      }
    })

    observe({
      if(input$exit > 0){
        stopApp()
      }
      
    })

    output$fhHistogram <- renderPlot({
      plot(rv$FH, init = TRUE, nls = TRUE, comps = TRUE)
    })

    output$flowNumber <- renderText({
      paste("File", tags$b(fhCurrent()), "of", length(.fhList))
    })

    observe({
      dat <- gateData()
      maxY <- max(dat[,2][!is.nan(dat[,2]) & is.finite(dat[,2])],
                  na.rm = TRUE)
      updateSliderInput(session, "yrange", 
                        step = max(6, ceiling(log(maxY)),
                                   na.rm = TRUE )/20,
                        max = max(6, ceiling(log(maxY)), na.rm = TRUE),
                        value = 0, min = 0)
    })

    ##browser()                           
    gateData <- reactive({
      chan1 <- input$xcol
      chan2 <- input$ycol

      raw <<- exprs(fhRaw(.fhList[[fhCurrent()]]))
      if(input$yType == "Y/X"){
        yvals <- raw[, chan2] / raw[, chan1]
        yName <- paste(chan2, chan1, sep = "/")
      } else if(input$yType == "Y"){
        yvals <- raw[, chan2]
        yName <- chan2
      }
      df <- data.frame(x = raw[, chan1],
                       y = yvals)
      names(df) <- c(chan1, yName)
      df
    })

    ################
    ## Gate Plots ##
    ################

    gateMar <- c(5, 4, 1, 0)
    output$gatePlot <- renderPlot({
      forceChange <- input$setGate
      dat <- gateData()
      op = par(mar = gateMar)
      ## need to account for infinite values when setting plot ranges.
      ## Infinite values generated when the dat[, 1] contains 0 values. 
      maxY <- max(dat[!is.nan(dat[,2]) & is.finite(dat[,2]), 2],
                  na.rm = TRUE)
      plot(dat, ylim = c(0, exp(log(maxY) - input$yrange)),
           type = 'n')
      if(isGated(.fhList[[fhCurrent()]])){
        points(dat[!fhGate(.fhList[[fhCurrent()]]), ], pch = 16,
               col = "#05050510")
        points(dat[fhGate(.fhList[[fhCurrent()]]), ], pch = 16,
               col = "#80000010") 
      } else
        points(dat, pch = 16, col = "#05050510")

      par(op)
    })

    output$gateResiduals <- renderPlot({
      op = par(mar = gateMar)
      plotResid(fhHistPlot(), main = "Gate Residuals", sub = "")
      par(op)
    })

    output$gatedData <- renderPlot({
      op = par(mar = gateMar)
      plot(fhHistPlot(), nls = FALSE, init = FALSE, comps = FALSE,
           main = "")
      par(op)
    })
    
    fhHistPlot <- reactive({
      ## returns the gated data, but doesn't set it permanently!
      dat <- gateData()
      bp <- brushedPoints(dat, xvar = names(dat)[1], yvar = names(dat)[2],
                          input$gatePlot_brush, allRows = TRUE)$selected_
      setGate(.fhList[[fhCurrent()]], bp, refresh = FALSE)
    })

  }
  runApp(shinyApp(ui = ui, server = server))
  if(length(.fhList) == 1){
    ## if we started with one flowHist object, return one flowHist object
    return(.fhList[[1]])
  } else {
    ## we started with a list of flowHist objects, return a list of
    ## modified objects:
    return(.fhList)
  }
}

